### PR TITLE
fix: override cycles must grow to max of all samples

### PR DESCRIPTION
**Pull request by simbig**
- [x] Added automated tests
- [x] Documented for all relevant versions
- [x] Updated the changelog

**Changes**

- All OverrideCycle total counts on DataSection must grow to the Reads-Sections maximum value by adding the N-tag

### CODE REVIEW

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 1]
```diff
@@ -49,4 +56,58 @@ protected function generateDataHeaderByProperties(array $samplePropertiesOfFirst
...
+            $this->maxIndex1Cycles(),
+            $this->maxRead2Cycles(),
+            $this->maxIndex2Cycles()
+        );
+    }
+
+    public function maxRead1Cycles(): int
+    {
+        $max = $this->dataRows
+            ->max(fn (BclSample $dataRow) => $dataRow->overrideCycles->read1->sumCountOfAllCycles());
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:73]:
```suggestion
            ->max(fn (BclSample $dataRow): int => $dataRow
                ->overrideCycles
                ->read1
                ->sumCountOfAllCycles());
```

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 2]
```diff
@@ -49,4 +56,58 @@ protected function generateDataHeaderByProperties(array $samplePropertiesOfFirst
...
+
+        return $max;
+    }
+
+    public function maxRead2Cycles(): ?int
+    {
+        $max = $this->dataRows->max(
+            fn (BclSample $dataRow) => $dataRow->overrideCycles->read2 instanceof OverrideCycle
+                ? $dataRow->overrideCycles->read2->sumCountOfAllCycles()
+                : null
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:84]:
```suggestion
            fn (BclSample $dataRow): ?int => $dataRow->overrideCycles->read2 instanceof OverrideCycle
                ? $dataRow->overrideCycles->read2->sumCountOfAllCycles()
                : null
```

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 3]
```diff
@@ -49,4 +56,58 @@ protected function generateDataHeaderByProperties(array $samplePropertiesOfFirst
...
+
+        assert(is_int($max) || is_null($max));
+
+        return $max;
+    }
+
+    public function maxIndex1Cycles(): int
+    {
+        $index1Cycles = $this->dataRows
+            ->max(fn (BclSample $dataRow) => $dataRow->overrideCycles->index1->sumCountOfAllCycles());
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:95]:
```suggestion
            ->max(fn (BclSample $dataRow): int => $dataRow
                ->overrideCycles
                ->index1
                ->sumCountOfAllCycles());
```

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 4]
```diff
@@ -49,4 +56,58 @@ protected function generateDataHeaderByProperties(array $samplePropertiesOfFirst
...
+        assert(is_int($index1Cycles));
+
+        return $index1Cycles;
+    }
+
+    public function maxIndex2Cycles(): ?int
+    {
+        $index2Cycles = $this->dataRows->max(
+            fn (BclSample $dataRow) => $dataRow->overrideCycles->index2 instanceof OverrideCycle
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:104]:
```suggestion
            fn (BclSample $dataRow): ?int => $dataRow->overrideCycles->index2 instanceof OverrideCycle
```

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 5]
```diff
@@ -49,4 +56,58 @@ protected function generateDataHeaderByProperties(array $samplePropertiesOfFirst
...
+    }
+
+    public function maxRead2Cycles(): ?int
+    {
+        $max = $this->dataRows->max(
+            fn (BclSample $dataRow) => $dataRow->overrideCycles->read2 instanceof OverrideCycle
+                ? $dataRow->overrideCycles->read2->sumCountOfAllCycles()
+                : null
+        );
+
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:86]:
```suggestion
```

File: src/IlluminaSampleSheet/V2/BclConvert/DataSection.php [hunk 6]
```diff
@@ -49,4 +56,58 @@ protected function generateDataHeaderByProperties(array $samplePropertiesOfFirst
...
+    }
+
+    public function maxIndex2Cycles(): ?int
+    {
+        $index2Cycles = $this->dataRows->max(
+            fn (BclSample $dataRow) => $dataRow->overrideCycles->index2 instanceof OverrideCycle
+                ? $dataRow->overrideCycles->index2->sumCountOfAllCycles()
+                : null
+        );
+
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/DataSection.php:108]:
```suggestion
```

File: src/IlluminaSampleSheet/V2/BclConvert/OverrideCycle.php [hunk 7]
```diff
@@ -7,17 +7,36 @@ class OverrideCycle
...
+        $this->cycles = $cycles;
     }
 
-    public function toString(): string
+    /** @param bool|null $isSecondIndexAndForwardDirection null if it is not the secondIndex, true if it is the secondIndex and forwardDirection, false if it is the secondIndex and reverseDirection */
+    public function toString(int $fillUpToMax, ?bool $isSecondIndexAndForwardDirection): string
     {
-        return implode('', array_map(
+        $countOfAllCycleTypes = $this->sumCountOfAllCycles();
+        assert($countOfAllCycleTypes <= $fillUpToMax, 'The sum of all cycle types must be less than or equal to the fill up to max value.');
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/OverrideCycle.php:20]: Seems like this should actually throw at runtime

File: src/IlluminaSampleSheet/V2/BclConvert/OverrideCycle.php [hunk 8]
```diff
@@ -7,17 +7,36 @@ class OverrideCycle
...
+        $remainingCycles = 'N' . ($fillUpToMax - $countOfAllCycleTypes);
+
+        return (bool) $isSecondIndexAndForwardDirection
+            ? $remainingCycles . $rawOverrideCycle
+            : $rawOverrideCycle . $remainingCycles;
+    }
+
+    public function sumCountOfAllCycles(): int
+    {
+        return array_sum(array_map(fn (CycleTypeWithCount $cycleTypeWithCount): int => $cycleTypeWithCount->count, $this->cycles));
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/OverrideCycle.php:40]:
```suggestion
        return array_sum(
            array_map(
                fn (CycleTypeWithCount $cycleTypeWithCount): int => $cycleTypeWithCount->count,
                $this->cycles
            )
        );
```

File: src/IlluminaSampleSheet/V2/BclConvert/OverrideCycles.php [hunk 9]
```diff
@@ -14,24 +15,40 @@ class OverrideCycles
...
         $this->read1 = $this->makeOverrideCycle($read1);
         $this->index1 = $this->makeOverrideCycle($index1);
         $this->index2 = $index2 !== null ? $this->makeOverrideCycle($index2) : null;
         $this->read2 = $read2 !== null ? $this->makeOverrideCycle($read2) : null;
+        $this->dataSection = $dataSection;
     }
 
     public function toString(): string
     {
+        $secondIndexIsForward = HeaderSection::indexOrientation() === HeaderSection::INDEX_ORIENTATION_FORWARD;
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/BclConvert/OverrideCycles.php:31]: This is completely static

File: src/IlluminaSampleSheet/V2/HeaderSection.php [hunk 10]
```diff
@@ -53,4 +56,12 @@ public function convertSectionToString(): string
 
         return implode("\n", $headerLines) . "\n";
     }
+
+    public static function indexOrientation(): string
+    {
+        // only support the default IndexOrientation,Forward for now.
+        // Added method to explicitly display than this flag
+        // influences the Index2 in the OverrideCycles
```
Comment (spawnia) [inline src/IlluminaSampleSheet/V2/HeaderSection.php:64]:
```suggestion
        // Only support the default IndexOrientation (Forward) for now.
        // Added method to explicitly display that this flag influences Index2 in OverrideCycles.
```

File: src/IlluminaSampleSheet/V2/BclConvert/OverrideCycle.php (hunk 7 above)
Comment (simbig) [reply src/IlluminaSampleSheet/V2/BclConvert/OverrideCycle.php:20]: Currently, this isn't logically possible, so I'm adding a pure assertion for development purposes.

**Review by spawnia — APPROVED** [review]

**Pull request merged**